core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -188,12 +194,18 @@ private void addAndWaitIfFull(CompletableFuture future, List<CompletableFuture>
          expiredMortal = ExpiryHelper.isExpiredMortal(lifespan, entry.getCreated(), currentTime);
       }
       if (expiredMortal) {
-         handleLifespanExpireEntry(entry.getKey(), value, lifespan);
+         // If the entry as found expired in memory during a write, that means we already have the lock for the
+         // given entry. Thus we must skip locking and wait for the remove expired to complete before continuing
+         // forward with the operation to let the event be raised properly.
```

Comment:
Do we guarantee that expiration listeners are notified synchronously with the read/write that triggered the expiration, or does "the event be raised properly" mean something else?

Reply:
yes they are done in a serial manner, expiration event has to come before the write event.